### PR TITLE
*: support read commit isolation level

### PR DESCRIPTION
**Pull request by nolouch**
cc @disksing @siddontang @zhangjinpeng1987 
will add test later.

### CODE REVIEW

File: src/storage/mvcc/reader.rs [hunk 1]
```diff
@@ -176,21 +180,27 @@ impl<'a> MvccReader<'a> {
...
+                        } else {
+                            // There is a pending lock. Client should wait or clean it.
+                            return Err(Error::KeyIsLocked {
+                                key: try!(key.raw()),
+                                primary: lock.primary,
+                                ts: lock.ts,
+                                ttl: lock.ttl,
+                            });
+                        }
+                    }
```
Comment (disksing) [inline src/storage/mvcc/reader.rs]: if isolation_level == SI && lock.ts < ts && lock.type==Put or Delete, we need to return the uncommitted value.

**Comment by disksing** [comment]
Better to attach a simple test with it, you may refer to https://github.com/pingcap/tidb/pull/3535/files#diff-bef3f3f6b29f8cfb6f1eb758c4a2c5bbR286.

File: src/storage/mvcc/reader.rs [hunk 2]
```diff
@@ -176,23 +180,29 @@ impl<'a> MvccReader<'a> {
...
-                    ts = lock.ts - 1;
-                } else {
-                    // There is a pending lock. Client should wait or clean it.
-                    return Err(Error::KeyIsLocked {
-                        key: try!(key.raw()),
-                        primary: lock.primary,
-                        ts: lock.ts,
-                        ttl: lock.ttl,
-                    });
+        // if self.isolation_level == IsolationLevel::SI
```
Comment (disksing) [inline src/storage/mvcc/reader.rs]: Remove this line.

**Comment by siddontang** [comment]
PTAL @huachaohuang

**Comment by siddontang** [comment]
PTAL @AndreMouche

File: src/storage/mvcc/reader.rs [hunk 3]
```diff
@@ -176,23 +180,28 @@ impl<'a> MvccReader<'a> {
...
-                    return Err(Error::KeyIsLocked {
-                        key: try!(key.raw()),
-                        primary: lock.primary,
-                        ts: lock.ts,
-                        ttl: lock.ttl,
-                    });
+        match self.isolation_level {
+            IsolationLevel::SI => {
+                if let Some(lock) = try!(self.load_lock(key)) {
+                    if lock.ts <= ts {
```
Comment (AndreMouche) [inline src/storage/mvcc/reader.rs]: Could we reduce the number of conditions in each `{}` (reduce the number of `{}`), may be we could define an independent function to check the lock?

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

File: src/storage/mvcc/reader.rs [hunk 4]
```diff
@@ -194,6 +197,19 @@ impl<'a> MvccReader<'a> {
                 }
             }
         }
+        Ok(Some(ts))
+    }
+
+    pub fn get(&mut self, key: &Key, mut ts: u64) -> Result<Option<Value>> {
+        // Check for locks that signal concurrent writes.
+        match self.isolation_level {
+            IsolationLevel::SI => {
```
Comment (AndreMouche) [inline src/storage/mvcc/reader.rs]: I prefer to use `if  self.isolation_leve==  IsolationLevel::SI` instead of `match` here?

Comment (nolouch) [inline src/storage/mvcc/reader.rs]: I intend to clearly show the difference between SI and RC with `match`.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**